File: valid-mountain-array.py
class Solution:
    def validMountainArray(self, arr: List[int]) -> bool:
        """
        
        
        [1,2,3,4,3,2,5]
         s
           e
        
        we want to use two pointers check if it is increase
        if isnt we want to set is increasing = False and isDecreasing = true
        if is decreasing is true and the array increases again we want to return false
        
        
        
        """
        is_increasing = False
        is_decreasing = False
        for i in range(1, len(arr)):
            if arr[i-1] == arr[i]:
                return False
            if arr[i-1] > arr[i]:
                is_decreasing = True
            
            if arr[i-1] < arr[i]:
                is_increasing = True
                
            if arr[i-1] < arr[i] and is_decreasing == True:
                return False
            
        return is_increasing and is_decreasing

File: 2020-algorithms-studying/Arrays/ValidMountainArray.js
/**
 * @param {number[]} arr
 * @return {boolean}
 */
var validMountainArray = function(arr) {
    
    //First edgecase
    if(arr.length < 3){
        return false
    }
    let isIncreasing = true;
    /**
    
    To be a mountain the increases have to increase by at least one every time
    And the decreases have to decrease by one every time
    **/
    let currentState;
    if(arr[0]>arr[1]){
        currentState = "decrease"
    }else{
        currentState = "increase"
    }
    let timesDecreased = 0
    
    for(let i = 0;i<arr.length;i++){
        let next = arr[i+1];
        let curr = arr[i]
        if(currentState=== "increase"){
            if(curr === next) return false
            if(curr < next){
                // console.log(`Increasing ${curr},${next}`)
            }else{
                if(next !== undefined){
                currentState = "decrease"
                timesDecreased++         
                }

            }
        }
        if(currentState === "decrease"){
            if(curr === next) return false
            if(curr > next  ){
                // console.log(`Decreasing${curr},${next}`)

            }else{
                if(next !== undefined){
                      return false
 
                }
            }
        }
    }
    console.log(timesDecreased,"TIEMS DECREASE")
    if(timesDecreased === 0) return false
    return isIncreasing
};